### PR TITLE
Add ADAMW "Fixing Weight Decay Regularization in Adam"

### PR DESCRIPTION
**Pull request by jarvist**
I really don't know what I'm doing, so this code may be all wrong. 

The motivation behind this method is in this blog post:
http://www.fast.ai/2018/07/02/adam-weight-decay/
And the original paper:
https://arxiv.org/abs/1711.05101.pdf

I just tried to add the relevant bits in Algorithm 2 of the paper. It moves the application of decay to the parameter update. (i.e. it becomes a mixing parameter). 

The paper argues that the L2 regularisation (which is what I believe `invdecay` is doing) are not equivalent to proper weight decay, are not equivalent when you have adaptive gradients.

Just an hours fun hacking one morning - I tested it against a toy Char-RNN I found on my harddrive, and it seems to improve the learning rate epoch-by-epoch over ADAM with the same decay parameter.

### CODE REVIEW

**Comment by MikeInnes** [comment]
Looks good! There are some tests for the optimisers in the tests dir – just adding ADAMW to the array there should be enough to add a test for this.

I'd appreciate a quick thumbs up from anyone else who knows the mathematics and can quickly verify it.

**Comment by jarvist** [comment]
I certainly appreciate another set of eyes!
If the AdamW paper is correct, this implies that all adaptive gradient optimisers are typically incorrectly L2 regularised rather than actually weight decayed. 
As such, `invdecay` was not actually the intended mathematics of these methods.

**Comment by jarvist** [comment]
Well adding ADAMW does run in the tests!
I must admit I'm slightly at a loss as to what the optimise tests do; 
https://github.com/FluxML/Flux.jl/blob/ce88273880730990ef2e236b775b2080eca12f4a/test/optimise.jl#L4-L17
Try and make a 10x10 random (normally distributed) matrix Hermitatian, by minimising the mean-squared error of the matrix (and it's transpose) acting on a random vector? 
Is this a good proxy for optimisation? Can you make any statement on learning rate from this?

As ADAMW===ADAM when the weight-decay is zero, should probably try them for some finite value of this.

**Comment by CarloLucibello** [comment]
> Try and make a 10x10 random (normally distributed) matrix Hermitatian, by minimising the mean-squared error of the matrix (and it's transpose) acting on a random vector?
Is this a good proxy for optimisation? Can you make any statement on learning rate from this?

I think is the intent here is to test the optimizers in the simplest convex scenario, a quadratic function. Doing `w'w`  assures that the form is positive definite, so that you actually do have a minima. If I rember correctly, optimal learning rate in this case should be the Lipschitz coefficient of the gradient

**Comment by jarvist** [comment]
Thanks for the pointers Carlo. I had a quick play - it seems that adding weight decay / regularisation seems to reduce both Adam and AdamW in the simple quadratic convex optimisation. 
Maybe a slightly more challenging toy system might be worth while to add? Or a different loop with << 10^5 steps?

**Comment by MikeInnes** [comment]
I'll merge this for now. Could be good to add another test, but I'm not sure what'll be meaningful for this beyond making sure the equation is right.

**Comment by jarvist** [comment]
Grand! 
It would be nice to have a slightly more challenging (even if still 'toy') test for the various optimisation. I guess there must be some simple system / potential-energy surface where weight-decay will benefit?

**Comment by CarloLucibello** [comment]
the rosenbrock function would be a good candidate https://en.wikipedia.org/wiki/Rosenbrock_function
